src/main/java/org/jvnet/solaris/libzfs/LibZFS.java

```diff
@@ -39,17 +39,122 @@
+        String libzfs4j_envvar_value;
+        libzfs4j_envvar_value = System.getenv(libzfs4j_envvar_name);
+        if (libzfs4j_envvar_value == null || libzfs4j_envvar_value.equals("")) {
+            libzfs4j_envvar_value = System.getProperty(libzfs4j_envvar_name);
+            if (libzfs4j_envvar_value == null || libzfs4j_envvar_value.equals("")) {
+                libzfs4j_envvar_value = "<NULL>";
+            }
+        }
+        return libzfs4j_envvar_value;
+    }
+
+    private void libzfs4j_add_feature(String libzfs4j_envvar_name, String libzfs4j_usevar_value, String libzfs4j_envvar_value) {
+        /* Add the key=value, reporting on it in log (and reporting the
+         * original envvvar/property value) for debugging. */
+        /* Note: we do not slow down to replace-if-present logic, since
+         * by current implementation and intention, this is only used from
+         * libzfs4j_init_features() below and that one only runs once.
+         * This decision can be amended later, if ever needed. */
+        Entry<String, String> myEntry;
+
+        System.out.println("[LIBZFS4J-DEBUG]: Setting '" +
+            libzfs4j_envvar_name + "' = '" + libzfs4j_usevar_value +
+            "' (envvar value was '" + libzfs4j_envvar_value + "')");
+        myEntry = new SimpleEntry<String, String>(libzfs4j_envvar_name, libzfs4j_usevar_value);
+        libzfs_features.add(myEntry);
+    }
+
+    void libzfs4j_init_features() {
+        String libzfs4j_envvar_value;
+        String libzfs4j_envvar_name;
+        String libzfs4j_usevar_value;
+        String libzfs4j_default_abi; /* Cache the default while we make decisions */
+
+        if (libzfs_features.size() > 0)
+            return; /* already inited */
+
+        libzfs4j_envvar_name = "LIBZFS4J_ABI";
+        libzfs4j_envvar_value = libzfs4j_get_extsetting(libzfs4j_envvar_name);
+        if (libzfs4j_envvar_value.equals("legacy") || libzfs4j_envvar_value.equals("openzfs")) {
+            /* Currently we recognize two values; later it may be more like openzfs-YYYY */
+            libzfs4j_default_abi = libzfs4j_envvar_value;
+        } else {
+            /* Detect presence of e.g. feature flags routines == openzfs */
+            libzfs4j_default_abi = "legacy";
+        }
+        libzfs4j_usevar_value = libzfs4j_default_abi;
+        libzfs4j_add_feature(libzfs4j_envvar_name, libzfs4j_usevar_value, libzfs4j_envvar_value);
+
+        libzfs4j_envvar_name = "LIBZFS4J_ABI_zfs_iter_snapshots";
+        libzfs4j_envvar_value = libzfs4j_get_extsetting(libzfs4j_envvar_name);
+        if (libzfs4j_envvar_value.equals("legacy") || libzfs4j_envvar_value.equals("3")) {
+            libzfs4j_usevar_value = "legacy";
+        } else {
+            if (libzfs4j_envvar_value.equals("openzfs") || libzfs4j_envvar_value.equals("4")) {
+                libzfs4j_usevar_value = "openzfs";
+            } else {
+                libzfs4j_usevar_value = libzfs4j_default_abi;
```

Comment:
It's not really about unsupportedness - rather a fallback for bad or missing setting. We have to define some state, even if wrong - but still predictable and repeatable.

So these checks mean (see more context in response below) that overall we consider either an `openzfs` or `legacy` ZFS ABI, however vague this definition is. And then for each routine that we support signature-picking for (only zfs_snapshot_init in this PR) we can detect or configure a particular choice of signature to use, or fall back to this overall default.